sdk/core/azure-core/azure/core/tracing/decorator.py

```
@@ -83,36 +84,64 @@ def distributed_trace(
+
+            # If tracing is disabled globally and user didn't explicitly enable it, don't trace.
+            if not tracing_enabled and user_enabled is None:
                 return func(*args, **kwargs)
 
             # Merge span is parameter is set, but only if no explicit parent are passed
             if merge_span and not passed_in_parent:
                 return func(*args, **kwargs)
 
-            with change_context(passed_in_parent):
-                name = name_of_span or get_function_and_class_name(func, *args)
-                with span_impl_type(name=name, kind=kind) as span:
-                    for key, value in func_tracing_attributes.items():
-                        span.add_attribute(key, value)
+            # Assume this will be popped in DistributedTracingPolicy.
+            func_tracing_attributes = kwargs.get("tracing_attributes", tracing_attributes)
+            span_attributes = {**func_tracing_attributes, **tracing_options.get("attributes", {})}
+
+            span_impl_type = settings.tracing_implementation()
+
+            name = name_of_span or func.__qualname__
+            if span_impl_type:
+                with change_context(passed_in_parent):
+                    with span_impl_type(name=name, kind=kind) as span:
+                        for key, value in span_attributes.items():
+                            span.add_attribute(key, value)  # type: ignore
+                        return func(*args, **kwargs)
+            else:
+                method_tracer = (
+                    tracer_provider.get_tracer() if tracer_provider else default_tracer_provider.get_tracer()
+                )
+                if not method_tracer:
+                    return func(*args, **kwargs)
+                with method_tracer.start_span(
+                    name=name,
+                    kind=kind,
+                    attributes=span_attributes,
+                    record_exception=tracing_options.get("record_exception", True),
```

we record exceptions on logs, right? If so, we don't need to record them on spans, and I'd rather not provide this configuration option. 

> When an instrumented operation fails with an exception, instrumentation SHOULD record
this exception as a [span event](/docs/exceptions/exceptions-spans.md) or a [log record](/docs/exceptions/exceptions-logs.md).

from [here](https://github.com/open-telemetry/semantic-conventions/blob/4f77620fe731c10d40f7d50c543d4e5c73a46ebf/docs/general/recording-errors.md?plain=1#L86-L87)

---

I don't believe we necessarily log exceptions. From what I see, most SDK client methods just raise `HttpResponseError` or some other `AzureError` without it being logged. By default, `record_exception` is True when starting a span. I can remove the configuration, but keep the default of OTel recording exceptions on the span. WDYT?

---

do we record exceptions on spans today? It seems not - https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-python%20record_exception&type=code

Can we avoid doing it? There is a high chance that span events will be deprecated in a few years and we'd regret using them.

---

Yep, we do record exceptions as span events today with the plugin. Exceptions are automatically recorded on method spans when the span's context is exited. This is the default behavior based on the OTel [start_span](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html#opentelemetry.trace.Tracer.start_span) defaults, so we haven't been explicitly calling `record_exception(...)`. Here, I was just maintaining the current behavior while providing some level of configuration to not record exception events. 

We could potentially turn this off by default, as I know some Azure Monitor OTel distro users in the past have complained that exceptions that they are handling in their code are still being logged in App Insights when they don't expect them to be. 

If a certain exception isn't in the span events list, and a user doesn't handle this exception, will App Insights get it through other means? @lzchen, perhaps you'd know?

---

If exception is not handled by the user code, the top-level instrumentation like django or flask would record it. Not sure if there is some global catch-all handler.

We don't record exceptions in Java and .NET and never heard complains about it (we do usually record them as logs though) 🤷‍♀️

---

> If exception is not handled by the user code, the top-level instrumentation like django or flask would record it. Not sure if there is some global catch-all handler.

I believe some of our instrumentations in Python have this behavior but some do not (especially the non-HTTP ones). Ideally we have exception handling handled consistently across all instrumentations.

> If a certain exception isn't in the span events list, and a user doesn't handle this exception, will App Insights get it through other means?

Exceptions not in span events can be logged (manually) through the logging SDK and then will be exported to AppInsights.

---

For the sake of avoiding duplicate exception info, and avoiding exception noise from being propagated to app insights when end users handle SDK exceptions, I'm going to disable recording exceptions as span events for these internal spans. Status will still be error and the `error.type` attribute will be populated with the exception name. If the top-level instrumentation library doesn't record the exception info, then users can choose to manually record the exception or log it. We'll see how that goes.